src/inet/routing/leach/Leach.cc

```diff
@@ -0,0 +1,450 @@
+//
+// SPDX-License-Identifier: LGPL-3.0-or-later
+//
+
+#include "inet/common/IProtocolRegistrationListener.h"
+#include "inet/common/ModuleAccess.h"
+#include "inet/common/ProtocolTag_m.h"
+#include "inet/linklayer/common/InterfaceTag_m.h"
+#include "inet/networklayer/common/L3AddressTag_m.h"
+#include "inet/physicallayer/wireless/common/contract/packetlevel/SignalTag_m.h"
+#include "inet/networklayer/common/L3AddressResolver.h"
+#include "inet/common/lifecycle/LifecycleOperation.h"
+#include "inet/common/lifecycle/ModuleOperations.h"
+#include "inet/common/lifecycle/NodeStatus.h"
+#include "inet/common/lifecycle/LifecycleController.h"
+#include "inet/routing/leach/Leach.h"
+
+namespace inet {
+namespace leach {
+
+Register_Enum(inet::leach::Leach, (Leach::ch, Leach::nch));
+Define_Module(Leach);
+
+Leach::Leach() {
+
+}
+
+Leach::~Leach() {
+    cancelAndDelete(event);
+    stop();
+}
+
+void Leach::initialize(int stage) {
+    RoutingProtocolBase::initialize(stage);
+
+    //reads from omnetpp.ini
+    if (stage == INITSTAGE_LOCAL) {
+        sequencenumber = 0;
+        host = getContainingNode(this);
+        interfaceTable.reference(this, "interfaceTableModule", true);
+
+        clusterHeadPercentage = par("clusterHeadPercentage");
+        numNodes = par("numNodes");
+
+        roundDuration = dblrand(0) * 10;
+        TDMADelayCounter = 1;
+        event = new cMessage("event");
+        roundDuration = dblrand(0) * 10;
+        wasCH = false;
+
+        vector<nodeMemoryObject> nodeMemory(numNodes); // Localized NCH node memory with CH data
+        vector<TDMAScheduleEntry> nodeCHMemory(numNodes); // Localized CH memory
+        vector<TDMAScheduleEntry> extractedTDMASchedule(numNodes); // Localized NCH node memory with extracted TDMA data
+
+    } else if (stage == INITSTAGE_ROUTING_PROTOCOLS) {
+        registerService(Protocol::manet, nullptr, gate("ipIn"));
+        registerProtocol(Protocol::manet, gate("ipOut"), nullptr);
+    }
+}
+
+void Leach::start() {
+    // Search the 802154 interface
+    configureInterfaces();
+
+    // schedules a random periodic event
+    event->setKind(SELF);
+    scheduleAt(simTime() + uniform(0.0, par("maxVariance").doubleValue()),
+            event);
+}
+
+void Leach::stop() {
+    nodeMemory.clear();
+    nodeCHMemory.clear();
+    extractedTDMASchedule.clear();
+    TDMADelayCounter = 1;
+}
+
+void Leach::configureInterfaces() {
+    int numInterfaces = interfaceTable->getNumInterfaces();
```

Comment:
This code picks the first wireless interface from the interface table. This must be documented in the NED file (i.e. that only the first can be used). I would rather go with pattern we use in other similar cases where you have an `interfaces` or `interface` string parameter and the user MUST explicitly specify which interface should be used by the module (you can use the cPatterMatcher similarly to the other MANET routing protocols)

Let's assume there is a host which has both WiFi and a 802.15.4 radio and WiFi happens to be on wlan0 and the sensor is on wlan1. with this code it is not possible to conigure wlan1 for Leach because wlan0 would be picked.

Reply:
Understood. Resolved in https://github.com/inet-framework/inet/pull/929/commits/ad80f49543c129b25da92bfee6b18b9a49c2a0d7